### PR TITLE
[opcua] Implementation of client side security policy

### LINKED ISSUE #595: OPC UA Driver should support client certificates
For an application we need to support Cert based Security in the OPC UA Driver.

Imported from Jira [PLC4X-185](https://issues.apache.org/jira/browse/PLC4X-185). Original Jira may contain additional context.
Reported by: jfeinauer.

### LINKED ISSUE #621: OPCUA response code and details are not reported in PLC response code
the attached screenshot shows an OPCUA response code as it is received by the underlying milo client, but the OPCUA driver reports only NOT_FOUND,
it would be great, if the underlying details are included in the OPCUA response code; the real cause for a problem is important for problem solution

Imported from Jira [PLC4X-304](https://issues.apache.org/jira/browse/PLC4X-304). Original Jira may contain additional context.
Reported by: torsten.uhr@sql-ag.de.

### LINKED ISSUE #625: Class casts in new opcua driver
Issue reported on mailing lists:

```

[nioEventLoopGroup-2-1] WARN io.netty.channel.DefaultChannelPipeline - An exceptionCaught() event was
fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline
did not handle the exception.
io.netty.handler.codec.DecoderException: java.lang.ClassCastException:
class org.apache.plc4x.java.opcua.readwrite.ServiceFault cannot be cast to class org.apache.plc4x.java.opcua.readwrite.ReadResponse
(org.apache.plc4x.java.opcua.readwrite.ServiceFault and org.apache.plc4x.java.opcua.readwrite.ReadResponse
are in unnamed module of loader 'app')
        at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:98)

       at io.netty.handler.codec.MessageToMessageCodec.channelRead(MessageToMessageCodec.java:111)

       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)

       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)

       at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)

       at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
 
      at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
      
 at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)

       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)

       at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)

       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)

       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

       at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)

       at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)

       at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)

       at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)

       at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)

       at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)

       at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassCastException:
class org.apache.plc4x.java.opcua.readwrite.ServiceFault cannot be cast to class org.apache.plc4x.java.opcua.readwrite.ReadResponse
(org.apache.plc4x.java.opcua.readwrite.ServiceFault and org.apache.plc4x.java.opcua.readwrite.ReadResponse
are in unnamed module of loader 'app')
        at org.apache.plc4x.java.opcua.protocol.OpcuaProtocolLogic.lambda$read$0(OpcuaProtocolLogic.java:177)

       at org.apache.plc4x.java.opcua.context.SecureChannel.lambda$4(SecureChannel.java:212)
     
  at org.apache.plc4x.java.spi.Plc4xNettyWrapper.decode(Plc4xNettyWrapper.java:175)
        at io.netty.handler.codec.MessageToMessageCodec$2.decode(MessageToMessageCodec.java:81)

       at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)

       ... 23 more

```

Imported from Jira [PLC4X-319](https://issues.apache.org/jira/browse/PLC4X-319). Original Jira may contain additional context.
Reported by: ldywicki.

### LINKED ISSUE #642: OPC UA silently fails to handle subscription response
While test of OPC UA simulator coming from freeopcua I found out that our client doesn't work with it. Maybe it is an issue of how they encode payload, maybe it is inconsistency in the spec, not sure - the end result is that actual response processing error gets swallowed and driver hangs.

Guilty code is in `OpcuaSubscriptionHandle`:
```

for (MonitoredItemCreateResult result : responseMessage.getResults().toArray(new MonitoredItemCreateResult[0]))
{
    if (OpcuaStatusCode.enumForValue(result.getStatusCode().getStatusCode()) != OpcuaStatusCode.Good)
{
        LOGGER.error("Invalid Field {}, subscription created without this field", fieldNames.get((int)
result.getMonitoredItemId()));
    } else {
        LOGGER.debug("Field {} was added to the subscription",
fieldNames.get((int) result.getMonitoredItemId() - 1)); // !monitored item id might be inconsistent!

   }
}

```

Copy of traffic causing issue in attachment.

Imported from Jira [PLC4X-342](https://issues.apache.org/jira/browse/PLC4X-342). Original Jira may contain additional context.
Reported by: ldywicki.

### PR DESCRIPTION
**Pull request by PatrykGala**
no text TBD

---

fixes https://github.com/apache/plc4x/issues/625
fixes https://github.com/apache/plc4x/issues/642
fixes https://github.com/apache/plc4x/issues/621
fixes https://github.com/apache/plc4x/issues/595

### CODE REVIEW

**Comment by chrisdutz** [comment]
Hi,

do you need any help with your PR? Would also be awesome, if you could sign up to our mailinglist and get a bit more in touch with the rest of the community dev@plc4x.apache.org (subscribe by sending an empty email to dev-subscribe@plc4x.apache.org).

Chris

**Comment by splatch** [comment]
@takraj Can you reserve some time next week to test this PR with your environment? Subscription tests are currently failing due to incomplete logic update, but these changes should address several aspects, including threading within driver.

**Comment by splatch** [comment]
Hey folks, could you have a look on these changes? I am working on last fixes for windows unit tests (I'll probably remove `certDirectory` option), all changes needed to get driver working with various security modes are in place ready for review.

File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/generation/WriteBufferByteBased.java
Comment (chrisdutz) [inline plc4j/spi/src/main/java/org/apache/plc4x/java/spi/generation/WriteBufferByteBased.java]: Admittedly I don#t quite understand neither the old nor the new version of this ;-)

**Review by chrisdutz — APPROVED** [review]
Had a look at mainly the changes in the shared parts (SPI) no objections. 
Didn't review the OPC-UA stuff, as you know I usually don't care about OPC-UA ;-)
But thanks for your work on it :-)

**Review by sruehl — APPROVED** [review]

**Review by hutcheb — APPROVED** [review]
Nothing holding you out from merging this.
Had a quick look, the existing functionality still seems to be working.
I haven't tested the security stuff yet, but will do eventually (Although a short readme on how to generate a keystore wouldn't go astray ;) )

**Comment by sruehl** [comment]
@splatch I would recommend the PR then ASAP otherwise we get into merge issues again at some point :)

**Comment by splatch** [comment]
Fair enough, I'll first rebase and then get a fresh sign off from @PatrykGala. For some reason github looses commit author when rebase is being made.

**Comment by chrisdutz** [comment]
Well ... you could merge in changes? Doesn't need to be a rebase ... especially when we do a squash merge back, right?

**Comment by splatch** [comment]
Rebase is not an issue. I think hardest part was getting it after all fixes we had prior 0.11 release.

I'd prefer to keep these four commits as they are, simply because first one is made by someone else (maybe we can squash failing unit test). Then there is mspec update which, for ease of porting to other languages, I'd leave alone too. Last commit is update of driver logic and adjustment of encryption logic which is large enough on its own to be left alone.

**Comment by chrisdutz** [comment]
The one thing I'm a bit worried about is that the mspec changes seem to imply a completely different structure ... so I'm a bit worried about which ones are actually the right ones? Were the changes only affecting message types that haven't been used in the past (Possibly still not being used)? If they were not used in the past, I guess updating the mspec doesn't cause any harm. If they have been used, I wonder which is the right format. One option would be to create a new "version" of the mspec (You know that we can provide multiple versions of an mspec and simply reference the version we want in the code-generator. So technically we could reference version 2 in java and version 1 in go.

**Comment by chrisdutz** [comment]
Had a look at the stuff outside of the OPC-UA driver directories and have no objections to merging this.

**Comment by splatch** [comment]
@chrisdutz I saw the dataio updates and I strive to merge this before you do. ;-) I will rebase it on top of current develop and wait for Patryk to amend sign-off on his commit. Today we will merge that, despite of remaining stability issue mentioned in #1364.

**Comment by sruehl** [comment]
Even if you would get a conflict with the generated code, it should be now problem as it always can be solved by regenerating :)

File: plc4j/drivers/opcua/src/test/resources/logback-test.xml
Comment (sruehl) [inline plc4j/drivers/opcua/src/test/resources/logback-test.xml]: check if we set this to info again before we merge.

File: plc4j/drivers/opcua/src/test/resources/log4j.properties
Comment (sruehl) [inline plc4j/drivers/opcua/src/test/resources/log4j.properties]: check if we set this to error again before we merge.

File: plc4go/internal/opcua/EncryptionHandler.go
Comment (sruehl) [inline plc4go/internal/opcua/EncryptionHandler.go]: is it ensured that this is always `readWriteModel.BinaryPayloadExactly`?. Otherwise this part will panic

File: plc4go/internal/opcua/EncryptionHandler.go
Comment (sruehl) [inline plc4go/internal/opcua/EncryptionHandler.go]: is it ensured that this is always `readWriteModel.BinaryPayloadExactly`?. Otherwise this part will panic

File: plc4go/internal/opcua/SecureChannel.go
Comment (sruehl) [inline plc4go/internal/opcua/SecureChannel.go]: is it ensured that this is always `readWriteModel.BinaryPayloadExactly`?. Otherwise this part will panic

File: plc4go/internal/opcua/SecureChannel.go
Comment (sruehl) [inline plc4go/internal/opcua/SecureChannel.go]: I think you can just use `.Stringer("chunk", opcuaResponse.GetChunk())`

File: plc4go/internal/opcua/SecureChannel.go
Comment (sruehl) [inline plc4go/internal/opcua/SecureChannel.go]: is it ensured that this is always `readWriteModel.BinaryPayloadExactly`?. Otherwise this part will panic

File: plc4go/internal/opcua/SecureChannel.go
Comment (sruehl) [inline plc4go/internal/opcua/SecureChannel.go]: is it ensured that this is always `readWriteModel.BinaryPayloadExactly`?. Otherwise this part will panic

File: plc4go/internal/opcua/SecureChannel.go
Comment (sruehl) [inline plc4go/internal/opcua/SecureChannel.go]: is it ensured that this is always `readWriteModel.BinaryPayloadExactly`?. Otherwise this part will panic

File: plc4go/internal/opcua/EncryptionHandler.go
Comment (splatch) [reply plc4go/internal/opcua/EncryptionHandler.go]: There is only one case when we expect extensible payload while parsing and it is `CloseRequest` (which we construct). For received data we always parse `BinaryPayload` cause this part of message might be segmented.
The extensible payload is more of a helper for generating payloads at runtime rather than for receiving them.

File: plc4j/drivers/opcua/src/test/resources/log4j.properties
Comment (splatch) [reply plc4j/drivers/opcua/src/test/resources/log4j.properties]: I removed log4j.properties and left only logback.

File: plc4j/drivers/opcua/src/test/resources/logback-test.xml
Comment (splatch) [reply plc4j/drivers/opcua/src/test/resources/logback-test.xml]: Updated to error.

**Comment by splatch** [comment]
Pushed on develop as:
* 6b033d3f7d29b8a20c716eea00d11acc2bbe70c1 (golang fixes)
* f8b0b3bf7782f14915d05145cbf2a621185fcd76 (checking of generated golang code)
* d0893d7ac786405783e0190676ba11e082511617 (chunking / encryption implementation)
* 6692f6f8eaac07f896bea8fef5b4b287c116c12e (mspec updates)
* 7202ec6530e09aa68f3106bbe26d45799d0a9c9a (test fixes)
* c572e58db5df605ad7b1564f528f88f85c0fa43a (implementation of Basic256 policy from @PatrykGala)

**Pull request closed without merging**
